### PR TITLE
Migrate relations

### PR DESCRIPTION
**Pull request by howbazaar**
Export and import relations.

### CODE REVIEW

File: state/migration/interfaces.go [hunk 1]
```diff
@@ -207,3 +210,36 @@ type Unit interface {
...
+// or a peer relation between different instances of a service.
+type Relation interface {
+	Id() int
+	Key() string
+
+	Endpoints() []Endpoint
+	AddEndpoint(EndpointArgs) Endpoint
+}
+
+// Endpoint represents one end of a relation. A named end-point provided
```
Comment (mjs) [inline state/migration/interfaces.go]: "A named endpoint is provided..." perhaps?

File: state/migration/model.go [hunk 2]
```diff
@@ -29,13 +30,14 @@ type ModelArgs struct {
 func NewModel(args ModelArgs) Model {
 	m := &model{
 		Version:             1,
-		Owner_:              args.Owner.Canonical(),
+		Owner_:              args.Owner.Id(),
```
Comment (mjs) [inline state/migration/model.go]: Why this change?

File: state/migration/model.go [hunk 3]
```diff
@@ -179,18 +188,82 @@ func (m *model) setServices(serviceList []*service) {
...
+	// Make sure that all the unit names specified in machine opened ports
+	// exist as units of services.
+	unknownUnitsWithPorts := unitsWithOpenPorts.Difference(allUnits)
+	if len(unknownUnitsWithPorts) > 0 {
+		return errors.Errorf("unknown unit names in open ports: %s", unknownUnitsWithPorts.SortedValues())
 	}
+
+	// Make sure that for each endpoint in each relation has settings for all
+	// units of that service.
+	for _, relation := range m.Relations_.Relations_ {
```
Comment (mjs) [inline state/migration/model.go]: This is big and ugly enough that perhaps it should be in it's own method.

File: state/migration/model_test.go [hunk 4]
```diff
@@ -124,7 +130,165 @@ func (*ModelSerializationSuite) TestParsingOptionals(c *gc.C) {
...
+}
+
+func (s *ModelSerializationSuite) addServiceToModel(model Model, name string, numUnits int) Service {
+	service := model.AddService(ServiceArgs{
+		Tag:                names.NewServiceTag(name),
+		Settings:           map[string]interface{}{},
+		LeadershipSettings: map[string]interface{}{},
+	})
+	service.SetStatus(minimalStatusArgs())
+	machineId := 0
```
Comment (mjs) [inline state/migration/model_test.go]: `machineId` doesn't appear to get used. Even if it was, won't `i` and `machineId` have the same value?

File: state/migration_import.go [hunk 5]
```diff
@@ -87,6 +90,9 @@ type importer struct {
 	dbModel *Model
 	model   migration.Model
 	logger  loggo.Logger
+	// serviceUnits is populate at the end of loading the services, and is a map
```
Comment (mjs) [inline state/migration_import.go]: populated

**Comment by mjs** [comment]
Ship It (with some minor fixes)

File: state/migration/model.go (hunk 2 above)
Comment (howbazaar) [inline state/migration/model.go]: Because an empty tag does not result in an empty Canonical strings, but "@local", which isn't useful.

I was wanting to test for an empty owner in the validation.

File: state/migration/interfaces.go (hunk 1 above)
Comment (howbazaar) [inline state/migration/interfaces.go]: Fixed.

File: state/migration_import.go (hunk 5 above)
Comment (howbazaar) [inline state/migration_import.go]: fixed

File: state/migration/model_test.go (hunk 4 above)
Comment (howbazaar) [inline state/migration/model_test.go]: Yeah... but it feels weird using the machineId as a unit number.

File: state/migration/model.go (hunk 3 above)
Comment (howbazaar) [inline state/migration/model.go]: OK.

**Comment by howbazaar** [comment]
$$merge$$
